CSharp/Library/Microsoft.Bot.Builder/Dialogs/IDialogContext.cs

```diff
@@ -151,6 +150,36 @@ public static partial class Extensions
             await botToUser.PostAsync(message, cancellationToken);
         }
 
+        /// <summary>
+        /// Ends the conversation.
+        /// </summary>
+        /// <param name="context">Communication channel to use.</param>
+        /// <param name="code">The conversation end code.</param>
+        /// <param name="message">Optional message to send to the user before ending the conversation</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns></returns>
+        public static async Task EndConversation(this IDialogContext context, string message = null, string code = EndOfConversationCodes.CompletedSuccessfully, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (!string.IsNullOrWhiteSpace(message))
+            {
+                // Send the desired message out to the user first
+                await context.PostAsync(message);
+            }
+
+            // Clear conversation data
+            context.ConversationData.Clear();
+            context.PrivateConversationData.Clear();
+
+            // Send EndOfConversation message
+            var eocMessage = context.MakeMessage();
+            eocMessage.Type = ActivityTypes.EndOfConversation;
+            eocMessage.AsEndOfConversationActivity().Code = code;
+
+            await context.PostAsync(eocMessage, cancellationToken);
+
+            // Clear the stack
+            context.Reset();

```

Comment:
hm. this code does (as far as i can tell) the same thing the Node SDK is doing so.... how is it possible/appropriate there but not here?

@Stevenic - any insight on the Node side?